src/NLog/Targets/FileTarget.cs

```diff
@@ -1759,50 +1623,85 @@ private byte[] GetLayoutBytes(Layout layout)
+            public const int CleanupPeriod = 2;
+
+            public const int MaxAllowed = 100;
+
+            public int Count
             {
-                if (this.recentAppenders[i] == null)
+                get
                 {
-                    break;
+                    return initializedFiles.Count;
                 }
+            }
+
+            public void Add(String fileName)
+            {
+                initializedFiles[fileName] = DateTime.Now;
+            }
+
+            public bool Remove(String fileName)
+            {
+                return initializedFiles.Remove(fileName);
+            }
 
-                if (this.recentAppenders[i].FileName == fileName)
+            public bool Contains(String fileName)
+            {
+                return initializedFiles.ContainsKey(fileName);
+            }
+
+            public IEnumerable<String> GetExpired(DateTime cleanupThreshold)
+            {
+                List<String> filesToUninitialize = new List<String>();
+
+                // Select the files require to be uninitialized.
+                foreach (var file in initializedFiles)
                 {
-                    this.recentAppenders[i].Close();
-                    for (int j = i; j < this.recentAppenders.Length - 1; ++j)
+                    if (file.Value < cleanupThreshold)
                     {
-                        this.recentAppenders[j] = this.recentAppenders[j + 1];
+                        filesToUninitialize.Add(file.Key);
                     }
-
-                    this.recentAppenders[this.recentAppenders.Length - 1] = null;
-                    break;
                 }
+
+                return filesToUninitialize;
             }
-        }
 
-#if !SILVERLIGHT
-        private static string CleanupInvalidFileNameChars(string fileName)
-        {
-            var lastDirSeparator =
-                fileName.LastIndexOfAny(new[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar });
+            public IEnumerable<String> GetItems()
+            {
+                return new List<String>(initializedFiles.Keys);
+            }
 
-            var fileName1 = fileName.Substring(lastDirSeparator + 1);
-            var dirName = lastDirSeparator > 0 ? fileName.Substring(0, lastDirSeparator) : string.Empty;
-            fileName1 = Path.GetInvalidFileNameChars().Aggregate(fileName1, (current, c) => current.Replace(c, '_'));
-            return Path.Combine(dirName, fileName1);
+            // Key = Filename, Value = Insterted Date/Time
+            private readonly Dictionary<string, DateTime> initializedFiles = new Dictionary<string, DateTime>();
         }
-#endif
 
-        private class DynamicFileArchive
+        private sealed class DynamicFileArchive
         {
-            public int MaxArchiveFileToKeep { get; set; }
+            /// <summary>
+            /// Max
+            /// </summary>
+            public int Size { get; set; }
```

Comment:
I am afraid Size cannot have a private setter because the value is assigned from MaxArchiveFiles from the containing class (FileTarget)